dwarf/main.c

```diff
@@ -0,0 +1,1011 @@
+			dwarf_dealloc (dbg, attr, DW_DLA_ATTR);
+			dwarf_dealloc (dbg, error, DW_DLA_ERROR);
+			return -1;//exit(1);
+		}
+
+		while (res != DW_DLV_NO_ENTRY) {
+			res = dwarf_tag (child, &tag, &error);
+			if (res != DW_DLV_OK) {
+				printf ("Error in getting tag for child of DW_TAG_array_type\n");
+				return -1;//exit(1);
+			}
+			strng *string1;
+			if (tag == DW_TAG_subrange_type) {
+				unsigned int arr_len;
+				res = get_array_length (dbg, child, &arr_len);
+				if (res == -1) {
+					string1 = strng_new("[]");
+				} else {
+					char tmparr[25];
+					sprintf (tmparr, "[%u]", arr_len);
+					string1 = strng_new (tmparr);
+				}
+				strngcat (p_name, *p_name, *string1, "");
+				strng_free (string1);
+			}
+			res = dwarf_siblingof (dbg, child, &sib, &error);
+			//dwarf_dealloc (dbg, child, DW_DLA_DIE);
+			child = sib;
+		}
+		//dwarf_dealloc (dbg, child, DW_DLA_DIE);
+		//dwarf_dealloc (dbg, sib, DW_DLA_DIE);
+		break;
+	}
+	case DW_TAG_subroutine_type: {
+		print_type (dbg, typedie, p_name);
+		Dwarf_Die child = 0;
+		Dwarf_Die sib = 0;
+		Dwarf_Half tag = 0;
+		res = dwarf_child (typedie, &child, &error);
+		if (res == DW_DLV_ERROR) {
+			printf ("Error in getting child of subroutine_type\n");
+			return -1;//exit(1);
+		}
+		if (res == DW_DLV_OK) {
+			strng *tmpstring = strng_new ("(");
+			strng *comstring = strng_new (",");
+			strngcat (p_name, *p_name, *tmpstring, " ");
+			strng_free (tmpstring);
+			while (res != DW_DLV_NO_ENTRY) {
+				strng *tmptype = strng_new ("");
+				res = dwarf_tag (child, &tag, &error);
+				if (res != DW_DLV_OK) {
+					printf ("Error in getting tag\n");
+					return -1;//exit(1);
+				}
+				if (tag == DW_TAG_formal_parameter) {
+					print_type (dbg, child, tmptype);
+					strngcat (p_name, *p_name, *tmptype, "");
+					strngcat (p_name, *p_name, *comstring, "");
+				}
+				res = dwarf_siblingof (dbg, child, &sib, &error);
+				if (res == DW_DLV_ERROR) {
+					printf ("Error in getting sibling of DW_AT_formal_parameter\n");
+					return -1;//exit(1);
+				}
+				dwarf_dealloc (dbg, child, DW_DLA_DIE);
+				child = sib;
+				strng_free (tmptype);
+			}
+			strng_free (comstring);
+			tmpstring = strng_new (")");
+			strngcat (p_name, *p_name, *tmpstring, "");
+			strng_free (tmpstring);
+		}
+		break;
+	}
+	default:
+		//printf ("\nNOT IMPLEMENTED.\n");
+		return tag;//exit(1); Return Tag
+	}
+	if (!localtype) {
+		dwarf_dealloc (dbg, typename, DW_DLA_STRING);
+	}
+	dwarf_dealloc (dbg, error, DW_DLA_ERROR);
+	dwarf_dealloc (dbg, typedie, DW_DLA_DIE);
+	dwarf_dealloc (dbg, attr, DW_DLA_ATTR);
+	return 0;
+}
+
+int get_type_size (Dwarf_Debug dbg, Dwarf_Die die, unsigned int *size) {
+	Dwarf_Attribute attr = 0;
+	Dwarf_Off offset;
+	Dwarf_Die typedie = 0;
+    Dwarf_Error error = 0;
+	Dwarf_Unsigned sz = 0;
+	int res = DW_DLV_ERROR;
+	Dwarf_Half tag = 0;
+
+	res = dwarf_attr (die, DW_AT_type, &attr, &error);
+	if (res == DW_DLV_ERROR) {
+		dwarf_dealloc (dbg, error, DW_DLA_ERROR);
+		exit(1);
```

Comment:
yes, I will remove all the exit. Currently using them so that I can be sure that code runs on the entire file.